File: packages/frontend/src/services/TaskService.js
// File: frontend/src/services/TaskService.js
import axios from 'axios';

const API_URL = '/api/tasks';

/**
 * Service for task-related API calls.
 */
class TaskService {
  /**
   * Get all tasks with pagination.
   * 
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getAllTasks(pageable = { page: 0, size: 20 }) {
    return axios.get(API_URL, { params: pageable });
  }

  /**
   * Get a task by ID.
   * 
   * @param {number} id - Task ID
   * @returns {Promise} - Promise with response data
   */
  getTaskById(id) {
    return axios.get(`${API_URL}/${id}`);
  }

  /**
   * Create a new task.
   * 
   * @param {Object} taskData - Task data
   * @returns {Promise} - Promise with response data
   */
  createTask(taskData) {
    return axios.post(API_URL, taskData);
  }

  /**
   * Update a task.
   * 
   * @param {number} id - Task ID
   * @param {Object} taskData - Updated task data
   * @returns {Promise} - Promise with response data
   */
  updateTask(id, taskData) {
    return axios.put(`${API_URL}/${id}`, taskData);
  }

  /**
   * Delete a task.
   * 
   * @param {number} id - Task ID
   * @returns {Promise} - Promise with response data
   */
  deleteTask(id) {
    return axios.delete(`${API_URL}/${id}`);
  }

  /**
   * Get tasks assigned to the current user.
   * 
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getMyTasks(pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/my-tasks`, { params: pageable });
  }

  /**
   * Get tasks assigned to a specific user.
   * 
   * @param {number} userId - User ID
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getTasksByAssignedUser(userId, pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/assigned-to/${userId}`, { params: pageable });
  }

  /**
   * Get tasks assigned to a specific team.
   * 
   * @param {number} teamId - Team ID
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getTasksByTeam(teamId, pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/team/${teamId}`, { params: pageable });
  }

  /**
   * Get tasks by status.
   * 
   * @param {string} status - Task status
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getTasksByStatus(status, pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/status/${status}`, { params: pageable });
  }

  /**
   * Get tasks by module.
   * 
   * @param {string} module - Module name
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getTasksByModule(module, pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/module/${module}`, { params: pageable });
  }

  /**
   * Get overdue tasks.
   * 
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getOverdueTasks(pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/overdue`, { params: pageable });
  }

  /**
   * Get tasks due today.
   * 
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getTasksDueToday(pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/due-today`, { params: pageable });
  }

  /**
   * Get upcoming tasks within a date range.
   * 
   * @param {Date} start - Start date
   * @param {Date} end - End date
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getUpcomingTasks(start, end, pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/upcoming`, {
      params: {
        start: start.toISOString(),
        end: end.toISOString(),
        ...pageable
      }
    });
  }

  /**
   * Get tasks related to a specific object.
   * 
   * @param {string} objectType - Object type
   * @param {number} objectId - Object ID
   * @returns {Promise} - Promise with response data
   */
  getTasksByRelatedObject(objectType, objectId) {
    return axios.get(`${API_URL}/related/${objectType}/${objectId}`);
  }

  /**
   * Get subtasks of a parent task.
   * 
   * @param {number} parentTaskId - Parent task ID
   * @returns {Promise} - Promise with response data
   */
  getSubtasks(parentTaskId) {
    return axios.get(`${API_URL}/${parentTaskId}/subtasks`);
  }

  /**
   * Assign a task to a user.
   * 
   * @param {number} taskId - Task ID
   * @param {number} userId - User ID
   * @returns {Promise} - Promise with response data
   */
  assignTaskToUser(taskId, userId) {
    return axios.put(`${API_URL}/${taskId}/assign-to-user/${userId}`);
  }

  /**
   * Assign a task to a team.
   * 
   * @param {number} taskId - Task ID
   * @param {number} teamId - Team ID
   * @returns {Promise} - Promise with response data
   */
  assignTaskToTeam(taskId, teamId) {
    return axios.put(`${API_URL}/${taskId}/assign-to-team/${teamId}`);
  }

  /**
   * Update the status of a task.
   * 
   * @param {number} taskId - Task ID
   * @param {string} status - New status
   * @returns {Promise} - Promise with response data
   */
  updateTaskStatus(taskId, status) {
    return axios.put(`${API_URL}/${taskId}/status/${status}`);
  }

  /**
   * Update the completion percentage of a task.
   * 
   * @param {number} taskId - Task ID
   * @param {number} percentage - Completion percentage
   * @returns {Promise} - Promise with response data
   */
  updateTaskCompletion(taskId, percentage) {
    return axios.put(`${API_URL}/${taskId}/completion/${percentage}`);
  }

  /**
   * Add a related object to a task.
   * 
   * @param {number} taskId - Task ID
   * @param {string} objectType - Object type
   * @param {number} objectId - Object ID
   * @param {string} relationshipType - Relationship type
   * @returns {Promise} - Promise with response data
   */
  addRelatedObjectToTask(taskId, objectType, objectId, relationshipType) {
    return axios.post(`${API_URL}/${taskId}/related-objects`, null, {
      params: {
        objectType,
        objectId,
        relationshipType
      }
    });
  }

  /**
   * Remove a related object from a task.
   * 
   * @param {number} taskId - Task ID
   * @param {number} relatedObjectId - Related object ID
   * @returns {Promise} - Promise with response data
   */
  removeRelatedObjectFromTask(taskId, relatedObjectId) {
    return axios.delete(`${API_URL}/${taskId}/related-objects/${relatedObjectId}`);
  }

  /**
   * Get unassigned tasks.
   * 
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  getUnassignedTasks(pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/unassigned`, { params: pageable });
  }

  /**
   * Search tasks.
   * 
   * @param {string} query - Search query
   * @param {Object} pageable - Pagination parameters
   * @returns {Promise} - Promise with response data
   */
  searchTasks(query, pageable = { page: 0, size: 20 }) {
    return axios.get(`${API_URL}/search`, {
      params: {
        query,
        ...pageable
      }
    });
  }
}

export default new TaskService();

File: packages/backend/src/main/java/com/enterprise/core/task/service/TaskService.java
// File: backend/src/main/java/com/enterprise/core/task/service/TaskService.java
package com.enterprise.core.task.service;

import com.enterprise.core.task.dto.TaskCreateDto;
import com.enterprise.core.task.dto.TaskDto;
import com.enterprise.core.task.entity.TaskModule;
import com.enterprise.core.task.entity.TaskStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Service interface for task management operations.
 */
public interface TaskService {

    /**
     * Create a new task.
     *
     * @param taskCreateDto The task creation data
     * @param createdById ID of the user creating the task
     * @return The created task
     */
    TaskDto createTask(TaskCreateDto taskCreateDto, Long createdById);

    /**
     * Get a task by ID.
     *
     * @param id The task ID
     * @return The task if found
     */
    TaskDto getTaskById(Long id);

    /**
     * Update a task.
     *
     * @param id The task ID
     * @param taskDto The updated task data
     * @return The updated task
     */
    TaskDto updateTask(Long id, TaskDto taskDto);

    /**
     * Delete a task.
     *
     * @param id The task ID
     */
    void deleteTask(Long id);

    /**
     * Get all tasks with pagination.
     *
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getAllTasks(Pageable pageable);

    /**
     * Get tasks assigned to a specific user.
     *
     * @param userId The user ID
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getTasksByAssignedUser(Long userId, Pageable pageable);

    /**
     * Get tasks assigned to a specific team.
     *
     * @param teamId The team ID
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getTasksByTeam(Long teamId, Pageable pageable);

    /**
     * Get tasks by status.
     *
     * @param status The task status
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getTasksByStatus(TaskStatus status, Pageable pageable);

    /**
     * Get tasks by module.
     *
     * @param module The module
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getTasksByModule(TaskModule module, Pageable pageable);

    /**
     * Get overdue tasks.
     *
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getOverdueTasks(Pageable pageable);

    /**
     * Get tasks due today.
     *
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getTasksDueToday(Pageable pageable);

    /**
     * Get upcoming tasks within a date range.
     *
     * @param start Start date
     * @param end End date
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getUpcomingTasks(LocalDateTime start, LocalDateTime end, Pageable pageable);

    /**
     * Get tasks related to a specific object.
     *
     * @param objectType The type of the related object
     * @param objectId The ID of the related object
     * @return List of tasks
     */
    List<TaskDto> getTasksByRelatedObject(String objectType, Long objectId);

    /**
     * Get subtasks of a parent task.
     *
     * @param parentTaskId The parent task ID
     * @return List of subtasks
     */
    List<TaskDto> getSubtasks(Long parentTaskId);

    /**
     * Assign a task to a user.
     *
     * @param taskId The task ID
     * @param userId The user ID
     * @return The updated task
     */
    TaskDto assignTaskToUser(Long taskId, Long userId);

    /**
     * Assign a task to a team.
     *
     * @param taskId The task ID
     * @param teamId The team ID
     * @return The updated task
     */
    TaskDto assignTaskToTeam(Long taskId, Long teamId);

    /**
     * Update the status of a task.
     *
     * @param taskId The task ID
     * @param status The new status
     * @return The updated task
     */
    TaskDto updateTaskStatus(Long taskId, TaskStatus status);

    /**
     * Update the completion percentage of a task.
     *
     * @param taskId The task ID
     * @param percentage The completion percentage
     * @return The updated task
     */
    TaskDto updateTaskCompletion(Long taskId, Integer percentage);

    /**
     * Add a related object to a task.
     *
     * @param taskId The task ID
     * @param objectType The type of the related object
     * @param objectId The ID of the related object
     * @param relationshipType The type of relationship
     * @return The updated task
     */
    TaskDto addRelatedObjectToTask(Long taskId, String objectType, Long objectId, String relationshipType);

    /**
     * Remove a related object from a task.
     *
     * @param taskId The task ID
     * @param relatedObjectId The related object ID
     * @return The updated task
     */
    TaskDto removeRelatedObjectFromTask(Long taskId, Long relatedObjectId);

    /**
     * Get unassigned tasks.
     *
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> getUnassignedTasks(Pageable pageable);

    /**
     * Search tasks.
     *
     * @param query The search query
     * @param pageable Pagination information
     * @return Page of tasks
     */
    Page<TaskDto> searchTasks(String query, Pageable pageable);
}